site/guides/variant-constructor-pattern.md

```diff
@@ -0,0 +1,182 @@
+namespace and name (aka GVKNN). If resource is cluster scoped, then the
+`metadata.name` uniquely identifies the resource in a cluster. If the resource
+is namespace scoped, then (namespace, name) together identifies the resource uniquely.
+
+[kpt-function-catalog](https://catalog.kpt.dev) provides two function that helps
+with customizing the identify of the resources:
+
+1. [set-namespace](https://catalog.kpt.dev/set-namespace/v0.3/): sets the
+   namespace for all resources in the package.
+2. [ensure-name-substring](https://catalog.kpt.dev/ensure-name-substring/v0.2/):
+   sets the name of the resources in the package.
+
+So you can use the appropriate functions from the catalog or implement a custom
+function to ensure unique identity of the resources.
+
+## Customizing non-identifier fields of resources
+
+Package can use other functions such as `set-labels`, `set-annotations`, `apply-replacements`
+or custom functions to transform other fields of resources.
+
+## Core mechanism
+
+Enabling automatic variant construction involves two steps:
+
+1. Use functions to customize identity or other fields of the resources
+2. Generating inputs for the functions declarared in the package
+
+Here is an example of `Kptfile` of a package that uses `set-namespace` and `apply-transform`
+to enable customization.
+
+```Kptfile
+# Kptfile
+...
+pipeline:
+  mutators:
+    - image: set-namespace:v0.3.4
+      configPath: ...
+    - image: apply-transform:v0.1.0
+      configPath: ...
+...
+```
+
+Now let's talk about the input to the functions. In most cases, variant's name
+(deployable package instance name) itself can be used to derive unique identity
+of the resources in the variant. For example, if I create a variant of
+`microservice` package, I will name the deployable package instance to
+`user-service` or `order-service`. So if the package's name is available to the
+functions, then they can use it to customize the name/namespace of the resources.
+So, starting with `kpt v1.0.0-beta.15+` , kpt makes `package's name` available
+in a local `ConfigMap` at a well-known path `package-context.yaml` that contains
+the package instance name and is available to functions during
+`kpt fn render` or `kpt fn eval`.
+
+Examples of `package-context.yaml` for abstract and deployable instance, For
+abstract package the value of `name` field is `example` and for deployable
+instance it contains actual value of the package name.
+
+```yaml
+# package-context.yaml
+# package context for an abstract package.
+# This is automatically created on `kpt pkg init`.
+apiVersion: v1
+kind: ConfigMap
+data:
+  name: example
+```
+
+```yaml
+# package-context.yaml
+# package context for a deployable instance of a package.
+# This is automatically populated during `kpt pkg get`.
+apiVersion: v1
+kind: ConfigMap
+data:
+  name: my-pkg-instance
+```
+
+kpt supports a way to create `deployable pkg instance` such that `package-context`
```

Comment:
Another place where we call it a different name, suggest sticking to one of `deployable package instance` or `deployable instance`